modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
+        Set<Index> affectedIndices = new HashSet<>();
+        Metadata metadata = state.metadata();
+        for (Index index : targetIndices) {
+            if (dataStream.isIndexManagedByDataStreamLifecycle(index, metadata::index)) {
+                IndexMetadata backingIndex = metadata.index(index);
+                assert backingIndex != null : "the data stream backing indices must exist";
+                List<DataStreamLifecycle.Downsampling.Round> downsamplingRounds = dataStream.getDownsamplingRoundsFor(
+                    index,
+                    metadata::index,
+                    nowSupplier
+                );
+                if (downsamplingRounds.isEmpty()) {
+                    continue;
+                }
+
+                String indexName = index.getName();
+                IndexMetadata.DownsampleTaskStatus backingIndexDownsamplingStatus = INDEX_DOWNSAMPLE_STATUS.get(backingIndex.getSettings());
+                String backingIndexDownsamplingSource = backingIndex.getSettings().get(IndexMetadata.INDEX_DOWNSAMPLE_SOURCE_NAME_KEY);
+
+                // if the current index is not a downsample we want to mark the index as read-only before proceeding with downsampling
+                if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource) == false
+                    && state.blocks().indexBlocked(ClusterBlockLevel.WRITE, indexName) == false) {
+                    // time to downsample, but first we have to mark the index as read-only
+                    affectedIndices.add(index);
+                    AddIndexBlockRequest addIndexBlockRequest = new AddIndexBlockRequest(WRITE, indexName).masterNodeTimeout(
+                        TimeValue.MAX_VALUE
+                    );
+                    transportActionsDeduplicator.executeOnce(
+                        addIndexBlockRequest,
+                        new ErrorRecordingActionListener(indexName, errorStore),
+                        (req, reqListener) -> addIndexBlock(addIndexBlockRequest, reqListener)
+                    );
+                    continue;
+                }
+
+                if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource)
+                    && backingIndexDownsamplingStatus.equals(SUCCESS)) {
+                    // if the backing index is a downsample index itself, let's check if its source index still exists as we must delete it
+                    Map<String, String> lifecycleMetadata = backingIndex.getCustomData(LIFECYCLE_CUSTOM_INDEX_METADATA_KEY);
+                    if (lifecycleMetadata == null || lifecycleMetadata.containsKey(REPLACEMENT_SOURCE_INDEX) == false) {
+                        // this is a downsampling index that was not added by data stream lifecycle in this data stream
+                        continue;
+                    }
+                    String actualDownsamplingSource = lifecycleMetadata.get(REPLACEMENT_SOURCE_INDEX);
```

Comment:
We need to use a custom to store the "true" source downsampling index because `IndexMetadata.INDEX_DOWNSAMPLE_SOURCE_NAME` does not in fact provide the source of the current downsample index, but the _first_ index that was first ever downsampled (it's kind of like `provided_name`, somehow, but not really - we might want to change this in the downsampling operation but that'll be something to followu-up on)

Note: this is significant for subsequent downsampling operations. Namely say we start with index `.ds-metrics-foo-2023.08.17-000001`. We first downsample it to `downsample-1s-.ds-metrics-foo-2023.08.17-000001`.
At this point `downsample-1s-.ds-metrics-foo-2023.08.17-000001` will have `INDEX_DOWNSAMPLE_SOURCE_NAME` configured to `.ds-metrics-foo-2023.08.17-000001`.

Now, any further downsample indices will continue to have `INDEX_DOWNSAMPLE_SOURCE_NAME` configured to `.ds-metrics-foo-2023.08.17-000001`.
e.g.  When downsampling `downsample-1s-.ds-metrics-foo-2023.08.17-000001` to `downsample-350s-.ds-metrics-foo-2023.08.17-000001`, the new index `downsample-350s-.ds-metrics-foo-2023.08.17-000001` will have `INDEX_DOWNSAMPLE_SOURCE_NAME` configured to `.ds-metrics-foo-2023.08.17-000001` (even though .ds-metrics-foo-2023.08.17-000001` is deleted at this point)